### PR TITLE
bpo-42627: Fix wrong parsing of Windows registry proxy settings

### LINKED ISSUE #2164: HTTP only system proxy not working properly on Windows
### Subject

Since Python 3.9, both `requests` and `pip` not working properly with a Windows system proxy set.

### Environment

Describe your environment.
At least, paste here the output of:

```python
>>> print("OS", platform.platform())
OS Windows-10-10.0.19041-SP0
>>> print("Python", platform.python_version())
Python 3.9.1
>>> print("urllib3", urllib3.__version__)
urllib3 1.26.2
```

```powershell
PS > Write-Output $env:https_proxy
PS > Write-Output $env:http_proxy
PS > Get-ItemProperty -Path 'HKCU:\Software\Microsoft\Windows\CurrentVersion\Internet Settings' | findstr ProxyServer
ProxyServer              : 127.0.0.1:7890
```

### Steps to Reproduce

1. Set Windows proxy from Setting - Network and Internet - Proxy. This proxy should support HTTP only. 
2. 
```python
import requests
requests.get("https://httpbin.org/robots.txt")
```

### Expected Behavior

Code executed without error. urllib and requests parse system proxy as HTTP proxy. 

### Actual Behavior

urllib parsed system proxy as **HTTPS** proxy and throw an error.
```
Traceback (most recent call last):
  File "C:\Python39\lib\site-packages\urllib3\connectionpool.py", line 696, in urlopen
    self._prepare_proxy(conn)
  File "C:\Python39\lib\site-packages\urllib3\connectionpool.py", line 964, in _prepare_proxy
    conn.connect()
  File "C:\Python39\lib\site-packages\urllib3\connection.py", line 359, in connect
    conn = self._connect_tls_proxy(hostname, conn)
  File "C:\Python39\lib\site-packages\urllib3\connection.py", line 496, in _connect_tls_proxy
    return ssl_wrap_socket(
  File "C:\Python39\lib\site-packages\urllib3\util\ssl_.py", line 432, in ssl_wrap_socket
    ssl_sock = _ssl_wrap_socket_impl(sock, context, tls_in_tls)
  File "C:\Python39\lib\site-packages\urllib3\util\ssl_.py", line 474, in _ssl_wrap_socket_impl
    return ssl_context.wrap_socket(sock)
  File "C:\Python39\lib\ssl.py", line 500, in wrap_socket
    return self.sslsocket_class._create(
  File "C:\Python39\lib\ssl.py", line 1040, in _create
    self.do_handshake()
  File "C:\Python39\lib\ssl.py", line 1309, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLEOFError: EOF occurred in violation of protocol (_ssl.c:1123)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Python39\lib\site-packages\requests\adapters.py", line 439, in send
    resp = conn.urlopen(
  File "C:\Python39\lib\site-packages\urllib3\connectionpool.py", line 755, in urlopen
    retries = retries.increment(
  File "C:\Python39\lib\site-packages\urllib3\util\retry.py", line 573, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='httpbin.org', port=443): Max retries exceeded with url: /robots.txt (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:1123)')))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Python39\lib\site-packages\requests\api.py", line 76, in get
    return request('get', url, params=params, **kwargs)
  File "C:\Python39\lib\site-packages\requests\api.py", line 61, in request
    return session.request(method=method, url=url, **kwargs)
  File "C:\Python39\lib\site-packages\requests\sessions.py", line 542, in request
    resp = self.send(prep, **send_kwargs)
  File "C:\Python39\lib\site-packages\requests\sessions.py", line 655, in send
    r = adapter.send(request, **kwargs)
  File "C:\Python39\lib\site-packages\requests\adapters.py", line 514, in send
    raise SSLError(e, request=request)
requests.exceptions.SSLError: HTTPSConnectionPool(host='httpbin.org', port=443): Max retries exceeded with url: /robots.txt (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:1123)')))
```
There are several issues around like https://github.com/psf/requests/issues/5740 and https://github.com/pypa/pip/issues/9216#issuecomment-741836058 , from those issues I was told it's a urllib issue.

### LINKED ISSUE #5740: Unable to use system proxy with HTTPS connection on Windows and Python 3.9
On Windows you can only set a "host:port" http proxy as system proxy.
```
PS > Get-ItemProperty -Path 'HKCU:\Software\Microsoft\Windows\CurrentVersion\Internet Settings' | findstr ProxyServer
ProxyServer              : 127.0.0.1:7890
```

But in Python, it assumes that this proxy supports both http and https on the same port:

```
>>> import urllib
>>> urllib.request.getproxies()
{'http': 'http://127.0.0.1:7890', 'https': 'https://127.0.0.1:7890', 'ftp': 'ftp://127.0.0.1:7890'}
```
Which will try to handshake httpps on http port. 

## Expected Result

On Python 3.8.x:

```
> .\python38.exe
Python 3.8.3 (tags/v3.8.3:6f8c832, May 13 2020, 22:37:02) [MSC v.1924 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import urllib, requests
>>> urllib.request.getproxies()
{'http': 'http://127.0.0.1:7890', 'https': 'https://127.0.0.1:7890', 'ftp': 'ftp://127.0.0.1:7890'}
>>> requests.get("http://www.google.com")
<Response [200]>
>>> requests.get("https://www.google.com")
<Response [200]>
```

## Actual Result

```
> python3
Python 3.9.1 (tags/v3.9.1:1e5d33e, Dec  7 2020, 17:08:21) [MSC v.1927 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import urllib, requests
>>> urllib.request.getproxies()
{'http': 'http://127.0.0.1:7890', 'https': 'https://127.0.0.1:7890', 'ftp': 'ftp://127.0.0.1:7890'}
>>> requests.get("http://www.google.com")
<Response [200]>
>>> requests.get("https://www.google.com")
Traceback (most recent call last):
  File "C:\Python39\lib\site-packages\urllib3\connectionpool.py", line 696, in urlopen
    self._prepare_proxy(conn)
  File "C:\Python39\lib\site-packages\urllib3\connectionpool.py", line 964, in _prepare_proxy
    conn.connect()
  File "C:\Python39\lib\site-packages\urllib3\connection.py", line 359, in connect
    conn = self._connect_tls_proxy(hostname, conn)
  File "C:\Python39\lib\site-packages\urllib3\connection.py", line 496, in _connect_tls_proxy
    return ssl_wrap_socket(
  File "C:\Python39\lib\site-packages\urllib3\util\ssl_.py", line 432, in ssl_wrap_socket
    ssl_sock = _ssl_wrap_socket_impl(sock, context, tls_in_tls)
  File "C:\Python39\lib\site-packages\urllib3\util\ssl_.py", line 474, in _ssl_wrap_socket_impl
    return ssl_context.wrap_socket(sock)
  File "C:\Python39\lib\ssl.py", line 500, in wrap_socket
    return self.sslsocket_class._create(
  File "C:\Python39\lib\ssl.py", line 1040, in _create
    self.do_handshake()
  File "C:\Python39\lib\ssl.py", line 1309, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLEOFError: EOF occurred in violation of protocol (_ssl.c:1123)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Python39\lib\site-packages\requests\adapters.py", line 439, in send
    resp = conn.urlopen(
  File "C:\Python39\lib\site-packages\urllib3\connectionpool.py", line 755, in urlopen
    retries = retries.increment(
  File "C:\Python39\lib\site-packages\urllib3\util\retry.py", line 573, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='www.google.com', port=443): Max retries exceeded with url: / (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:1123)')))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Python39\lib\site-packages\requests\api.py", line 76, in get
    return request('get', url, params=params, **kwargs)
  File "C:\Python39\lib\site-packages\requests\api.py", line 61, in request
    return session.request(method=method, url=url, **kwargs)
  File "C:\Python39\lib\site-packages\requests\sessions.py", line 542, in request
    resp = self.send(prep, **send_kwargs)
  File "C:\Python39\lib\site-packages\requests\sessions.py", line 655, in send
    r = adapter.send(request, **kwargs)
  File "C:\Python39\lib\site-packages\requests\adapters.py", line 514, in send
    raise SSLError(e, request=request)
requests.exceptions.SSLError: HTTPSConnectionPool(host='www.google.com', port=443): Max retries exceeded with url: / (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:1123)')))
```

## Reproduction Steps

Set a http Windows system proxy then execute:
```python
import requests
requests.get("https://www.google.com")
```

## System Information

    $ python -m requests.help

```
{
  "chardet": {
    "version": "4.0.0"
  },
  "cryptography": {
    "version": ""
  },
  "idna": {
    "version": "2.10"
  },
  "implementation": {
    "name": "CPython",
    "version": "3.9.1"
  },
  "platform": {
    "release": "10",
    "system": "Windows"
  },
  "pyOpenSSL": {
    "openssl_version": "",
    "version": null
  },
  "requests": {
    "version": "2.25.1"
  },
  "system_ssl": {
    "version": "1010107f"
  },
  "urllib3": {
    "version": "1.26.2"
  },
  "using_pyopenssl": false
}
```

### PR DESCRIPTION
**Pull request by CrazyBoyFeng**
In Windows registry proxy settings, `https=host:456` does not mean that the proxy server communicates with the client through the HTTPS protocol, but it means that the proxy server proxies the client's traffic to the HTTPS website. The proxy server itself communicates with the client through the HTTP protocol by default. 

fix pypa/pip#9568 fix pypa/pip#9614 fix urllib3/urllib3#2164 fix psf/requests#5740

https://bugs.python.org/issue42627

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).



## CLA Missing

Our records indicate the following people have not signed the CLA:

@CrazyBoyFeng

For legal reasons we need all the people listed to sign the CLA before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Comment by CrazyBoyFeng** [comment]
I do not know why the `Azure Pipelines PR — #20210522.11` check failed, I passed the test on my computer.

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

## Recognized GitHub username

We couldn't find a [bugs.python.org](https://bugs.python.org/) (b.p.o) account corresponding to the following GitHub usernames:

@testinguser

This might be simply due to a missing "GitHub Name" entry in one's b.p.o account settings. This is necessary for legal reasons before we can look at this contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.




You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

File: Lib/urllib/request.py [hunk 1]
```diff
@@ -2674,22 +2674,26 @@ def getproxies_registry():
...
+                    if not re.match('(?:[^/:]+)://', address):
+                        # Add type:// prefix to address without specifying type
+                        if protocol in ('http', 'https', 'ftp'):
+                            # The default proxy type of Windows is HTTP
+                            address = 'http://' + address
+                        elif protocol == 'socks':
+                            address = 'socks://' + address
+                    proxies[protocol] = address
+                # Use SOCKS proxy for HTTP(S) protocols
+                if proxies['socks']:
```
Comment (keuin) [inline Lib/urllib/request.py]: Line 2692 should be `if proxies.get('socks')`, otherwise it raises a `ValueError` in my environment.

Comment (CrazyBoyFeng) [reply Lib/urllib/request.py]: Sorry, I have fixed it.

File: Lib/urllib/request.py [hunk 2]
```diff
@@ -2674,22 +2674,26 @@ def getproxies_registry():
...
+                            address = 'http://' + address
+                        elif protocol == 'socks':
+                            address = 'socks://' + address
+                    proxies[protocol] = address
+                # Use SOCKS proxy for HTTP(S) protocols
+                if proxies.get('socks'):
+                    # The default SOCKS proxy type of Windows is SOCKS4
+                    address = re.sub(r'^socks://', 'socks4://', proxies['socks'])
+                    proxies['http'] = proxies['http'] or address
+                    proxies['https'] = proxies['https'] or address
```
Comment (zooba) [inline Lib/urllib/request.py]: Are we guaranteed that `proxies` already has `'http'` and `'https'` keys at this point? I don't see how, so probably better to use `.get()` here.

**Comment by zooba** [comment]
@CrazyBoyFeng Looks like you're going to hit a known error in CI right now. Merging the latest from master should fix it.

**Comment by CrazyBoyFeng** [comment]
I've synced the branch to the latest, hope this works.

**Comment by gfxonline** [comment]
Any updates on this PR?

**Comment by CrazyBoyFeng** [comment]
Until this bug is fixed, when using a proxy on Windows, you can work around it by:
* As a developer, use the monkey patch. Or manually read the registry and write the environment variable `https_proxy`.
* As an application user, write `https_proxy` to the environment variable.

**Comment by gfxonline** [comment]
> Until this bug is fixed, when using a proxy on Windows, you can work around it by:
> 
> * As a developer, use the monkey patch. Or manually read the registry and set the environment variable `https_proxy=http://proxy`.
> * As an application user, set `https_proxy=http://proxy` to the environment variable.
> 
> _If it is a socks proxy, you should set environment variables `http_proxy=socks4://proxy` and `https_proxy=socks4://proxy`._

That matches my envvars settings and I get the same problem. I was under the impression that this PR solves this exact issue where https is used to communicate with the proxy regardless of the protocol defined in the environment variables.

**Comment by sethmlarson** [comment]
Hello! I'm happy to see this PR, up until this point I didn't know that this issue existed which was causing a lot of confusion for urllib3 users. If I could suggest an alternate solution, instead of introducing any changes related to SOCKS we change the following line to be `http` instead of `https`:

```python
# Before:
proxies['https'] = 'http://%s' % proxyServer

# After:
proxies['https'] = 'http://%s' % proxyServer
```

And then keep the rest of the code as-is. My thought is this is the first mention of SOCKS in CPython as far as I can tell and adding support for SOCKS is a much bigger feature addition compared to the smaller fix of making HTTP and HTTPS requests use the same "default" proxy scheme.

**Comment by CrazyBoyFeng** [comment]
@sethmlarson 
Although the previous code did not explicitly handle socks proxies. But it can pass `http=socks://host:port` from the environment variable. And downstream libraries such as `requests` can receive and parse it correctly.
So this doesn't mean that cpython didn't support socks proxy before.
Even if it didn't support it before, it should support it now.

File: Misc/NEWS.d/next/Library/2021-05-22-07-58-59.bpo-42627.EejtD0.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Fix wrong parsing of Windows registry proxy settings
```
Comment (zooba) [inline Misc/NEWS.d/next/Library/2021-05-22-07-58-59.bpo-42627.EejtD0.rst]:
```suggestion
Fix incorrect parsing of Windows registry proxy settings
```

Comment (zooba) [reply Misc/NEWS.d/next/Library/2021-05-22-07-58-59.bpo-42627.EejtD0.rst]: Mainly trying to see whether this adds the newline needed.

**Comment by zooba** [comment]
Looks like if I edit the NEWS file through a suggestion it doesn't add the newline at the end of the file. We can't merge without it, so can you make sure that it has exactly one newline? You'll probably need to pull before changing, as I made a commit that won't merge cleanly.

**Comment by CrazyBoyFeng** [comment]
I added a <LF> at the end of file. It now passed the test.

**Comment by cyfex** [comment]
In Ubuntu 22.04, pip 23.1.2, python 3.10, this issue is still an issue. It seems not to be able to handle socks5 proxy properly.

**Comment by zooba** [comment]
Please file a new issue. It's no good commenting on an already completed PR that had nothing to do with your concern.

**Pull request merged**